cppfd/solver.cpp

```diff
@@ -129,91 +151,158 @@ void Solver::solve(stopping_point s_p, linear_solver l_s,adaptative_time_step at
   if(this->verbosity > 0){
-    std::cout<<std::endl;
-    std::cout<<"Done."<<std::endl;
-    std::cout<<"Total computation duration: " << time3 - time2 <<std::endl;
-    std::cout<<std::endl;
-    std::cout<<"Predictor step total duration: " << this->time_predictor_step <<" ("<<this->time_predictor_step * 100/ (time3 - time2)<<" %)"<<std::endl;
-    std::cout<<"RHS assembly total duration: " << this->time_assemble_RHS <<" ("<<this->time_assemble_RHS * 100/ (time3 - time2)<<" %)"<<std::endl;
-    std::cout<<"Linear solver for poisson equation total duration: " << this->time_poisson_solve <<" ("<<this->time_poisson_solve * 100/ (time3 - time2)<<" %)"<<std::endl;
-    std::cout<<"Corrector step total duration: " << this->time_corrector_step <<" ("<<this->time_corrector_step * 100/ (time3 - time2)<<" %)"<<std::endl;
+    std::cout << std::endl;
+    std::cout << "Done."
+	      << std::endl;
+    std::cout << "Total computation duration: "
+	      << time3 - time2
+	      << std::endl;
+    std::cout << std::endl;
+    std::cout << "Predictor step total duration: "
+	      << time_predictor_step
+	      << " ("
+	      << time_predictor_step * 100 / (time3 - time2)
+	      << " %)"
+	      << std::endl;
+    std::cout << "RHS assembly total duration: "
+	      << time_assemble_RHS
+	      << " ("
+	      << time_assemble_RHS * 100 / (time3 - time2)
+	      << " %)"
+	      << std::endl;
+    std::cout << "Linear solver for poisson equation total duration: "
+	      << time_poisson_solve
+	      << " ("
+	      << time_poisson_solve * 100 / (time3 - time2)
+	      << " %)"
+	      << std::endl;
+    std::cout << "Corrector step total duration: "
+	      << time_corrector_step
+	      << " ("
+	      << time_corrector_step * 100 / (time3 - time2)
+	      << " %)"
+	      << std::endl;
   }
   timer.reset();
 }
 
-// computation of laplacian matrix
-void Solver::assemble_laplacian(){
-  // initialize container for Laplacian
-  const int m = this->mesh.get_n_cells_x();
-  const int n = this->mesh.get_n_cells_y();
-  const int mn = m * n;
-  this->laplacian = Kokkos::View<double**>("laplacian", mn, mn);
-
-  for(int j = 0; j < n; j++){
-    for(int i = 0; i < m; i++){
-      int k = this->mesh.cartesian_to_index(i, j, m, n);
-      // initialize diagonal value
-      int v = -4;
-
-      // detect corners and borders
-      if(i == 0 || i == m - 1)
+// computation of Laplacian matrix
+double Solver::assemble_Laplacian(){
+  // initialize containers for sparse storage of Laplacian
+  const uint64_t m = this->mesh.get_n_cells_x();
+  const uint64_t n = this->mesh.get_n_cells_y();
+  const uint64_t mn = m * n;
+  const uint64_t nnz = 5 * mn - 2 * (m + n);
+  Kokkos::View<uint64_t*> row_ptrs("row pointers", mn + 1);
+  Kokkos::View<uint64_t*> col_ids("column indices", nnz);
+  Kokkos::View<double*> values("values", nnz);
+
+  // iterate m*n over cells to construct mn*mn Laplacian
+  uint64_t idx = 0;
+  for(uint64_t j = 0; j < n; j++){
```

Comment:
This might be easier to read as a separate function that produces the sparse matrix.

Reply:
@lifflander I made a refactoring following your comment, but also extended it in order to expand the use of the Kokkos arithmetic traits `zero` and `one`, can you please have a look? Thanks